app/domain/authentication/authn_jwt/authenticator.rb

```diff
@@ -0,0 +1,144 @@
+require 'command_class'
+
+module Authentication
+  module AuthnJwt
+    # Generic JWT authenticator that receive JWT vendor configuration and uses to validate that the authentication
+    # request is valid, and return conjur authn token accordingly
+    Authenticator = CommandClass.new(
+      dependencies: {
+        token_factory: TokenFactory.new,
+        logger: Rails.logger,
+        audit_log: ::Audit.logger,
+        validate_origin: ::Authentication::ValidateOrigin.new,
+        role_class: ::Role,
+        webservice_class: ::Authentication::Webservice,
+        validate_role_can_access_webservice: ::Authentication::Security::ValidateRoleCanAccessWebservice.new
+      },
+      inputs: %i[jwt_configuration authenticator_input]
+    ) do
+      extend(Forwardable)
+      def_delegators(:@authenticator_input, :account, :username, :client_ip, :authenticator_name, :service_id)
+
+      def call
+        create_authentication_parameters
+        validate_and_decode_token
+        get_jwt_identity
+        validate_user_has_access_to_webservice
+        validate_origin
+        validate_restrictions
+        audit_success
+        new_token
+      rescue => e
+        audit_failure(e)
+        raise e
+      end
+
+      private
+
+      def create_authentication_parameters
+        authentication_parameters
+      end
+
+      def authentication_parameters
+        @authentication_parameters ||= @jwt_configuration.authentication_parameters(@authenticator_input)
+      end
+
+      def validate_and_decode_token
+        @logger.debug(LogMessages::Authentication::AuthnJwt::VALIDATE_AND_DECODE_TOKEN.new)
+        @authentication_parameters.decoded_token = @jwt_configuration.validate_and_decode_token(@authentication_parameters)
+      end
+
+      def get_jwt_identity
+        # Will be changed when real get identity implemented
+        @logger.debug(LogMessages::Authentication::AuthnJwt::GET_JWT_IDENTITY.new)
+        @authentication_parameters.jwt_identity =  @jwt_configuration.jwt_identity(@authentication_parameters)
```

Comment:
i've said it before but this line of code really troubles me. we set a member of the `authentication_parameters` object that depends on the `authentication_parameters` object itself.